### ISSUE TITLE
Failed to create timer for background DNS resolution: too many pending timers

### ISSUE THREAD

**Issue by artificerpi** [Kong/kong#6742]
### Summary

I'm using Kong(`2.1.4`) cluster in CP/DP mode, including 2 CP pods (DNS HA via kong hosts file) and 5 DP pods. 

RPS is almost 1K, but configuration is updated frequently.

The cluster was running smoothly, but suddenly one pod ran into following error, and services routed to this pod responded 
`{“message": "An unexpected error occurred"}`  

Pretty wierd :<

### Steps To Reproduce

Don't know yet.

Using default cluster configuration but setting `kong_nginx_work_process=1` to get rid of [`dns resolution issues`](https://github.com/Kong/kong/issues/5455)

### Additional Details & Logs

- Kong version `2.1.4`
- Kong debug-level startup logs (Not enabled)
- Kong error logs

```
2021/01/18 14:31:57 [notice] 24102#0: *910 [lua] cache. lua:374: purge): [DB cache] purging (local) cache, context: ngx.timer 
2021/01/18 14:31:57 [notice] 24102#0: *910 [lua] cache.lua:374: purge): [DB cachel purging (local) cache, context: ngx.timer 
2021/01/18 14:31:57 [error] 24102#0: *910 [lua] timer.lua:106: resty_timer(): [resty-timer] failed to create timer: too many pending timers, context: ngx.timer
2021/01/18 14:31:57 [error] 24102#0: *910 [lua] handler.lua:452: config flip failed: / usr/local/share/lua/5.1 /resty/ dns/balancer/ring.lua:435: failed to create timer for background
DNS resolution: too many pending timers, Context: ngx.timer 
2021/01/18 14:31:57 [error 24102#0: *302966382 [lua] balancer.lua:706: get_balancer(): balancer not found for <SERVICE_NAME>, will create it, client: 10.55.193.136, server: kong, request: "GET / HTTP/
1.1", host: "DOMAIN",
2021/01/18 14:31:57 [error] 24102#0: *302966382 [lua] timer. lua:106: resty_timer(): [resty-timer] failed to create timer: too many pending timers, client: 10.55.193.136, server: kong, request: "GET / HTTP/
1.1", host: "DOMAIN"
2021/01/18 14:31:57 [error] 24102#0: *302966382 lua entry thread aborted: runtime error: / usr/local/share/lua/5.1 /resty/ dns/balancer/ring. lua:435: failed to create timer for background DNS resolution: too many pending timers
stack traceback:
coroutine
        [C]: in function 'assert'
                /usr/local/share/lua/5.1/resty/dns/balancer/ring.lua:435: in function 'new'
                /usr/local/share/lua/5.1/kong/runloop/balancer.lua:471: in function 'create_balancer_exclusive'
                /usr/local/share/lua/5.1/kong/runloop/balancer.lua:533: in function 'create_balancer'
                /usr/local/share/lua/5.1/kong/runloop/balancer.lua:707: in function 'get_balancer'
                /usr/local/share/lua/5.1/kong/runloop/balancer.lua:1080: in function 'execute'
                /usr/local/share/lua/5.1/kong/runloop/handler.lua:891: in function 'balancer_execute'
                /usr/local/share/lua/5.1/kong/runloop/handler.lua:1264: in function 'after'
                /usr/local/share/lua/5.1/kong/init.lua:788: in function 'access'
                access_by_lua(nginx-kong.conf:92):2: in main chunk, client: 10.55.193.136, server: kong, request: "GET / HTTP/1.1", host: "DOMAIN", referrer: "referrer_host"
```
- Kong configuration (the output of a GET request to Kong's Admin port - see
  https://docs.konghq.com/latest/admin-api/#retrieve-node-information)
- Operating system `Linux centos 7.6`

**Comment by bungle** [comment]
@artificerpi, can you please try 2.3.0?

**Comment by artificerpi** [comment]
Not so fast. I have to verify version 2.3.0 on staging environment before going into production. Any workarounds to fix it?

**Comment by artificerpi** [comment]
Added info. Spotted following repeated warning logs before running into error

```
2021/01/20 10:28:26 [warn] 209606#0: *913 [lua] ring.lua:246: redistributeIndices(): [upstream:srv-name 9710] redistributed indices, size=100, dropped=100, assigned=0, left unassigned=100, context: ngx.timer
2021/01/20 10:28:26 [notice] 209606#0: *913 [lua] cache.lua:374: purge(): [DB cache] purging (local) cache, context: ngx.timer
2021/01/20 10:28:26 [notice] 209606#0: *913 [lua] cache.lua:374: purge(): [DB cache] purging (local) cache, context: ngx.timer
```

The trigger operation is to update tags of services, routes and upstreams. why is that the dropped amount is 100 here ?

**Comment by locao** [comment]
> Any workarounds to fix it?

You may try to increase the max timers values like this:
```
lua_max_running_timers 32768;
lua_max_pending_timers 32768;
```

But the best approach to this is upgrading. Kong 2.3.0 uses new versions of lua-resty-dns-client and lua-resty-healthcheck that don't use individual timers by upstream, but a single timer for all of them, so you won't have issues with timers resources being exhausted.

**Comment by bungle** [comment]
@artificerpi, or wait for `2.3.1` if you are using healthchecks. Other issue was spotted with that and we are delivering the fix and release soon.

**Comment by guanlan** [comment]
This should be fixed in the recent released 2.3.1.

**Comment by Mayank-Sinha-Infrrd** [comment]
I am using kong 2.5.2 , getting the same error
